### PR TITLE
[net/http] enhance otelhttp example to support metric

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

The current otelhttp example only contains the logic of traces, I added the support of metric.

### CODE REVIEW

**Comment by dmathieu** [comment]
The metrics SDK is currently under heavy revamping to be moved out of beta (see the [new_sdk branch](https://github.com/open-telemetry/opentelemetry-go/compare/new_sdk/main)).
I think this kind of change should wait for that work to be finished.

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 1]
```diff
@@ -53,6 +59,32 @@ func initTracer() (*sdktrace.TracerProvider, error) {
...
+		controller.WithExporter(metricExporter),
+		controller.WithCollectPeriod(1*time.Second),
+	)
+
+	err = controller.Start(ctx)
+	if err != nil {
+
+	}
+	global.SetMeterProvider(controller)
+
```
Comment (hanyuancheung) [inline instrumentation/net/http/otelhttp/example/server/server.go:84]:
```suggestion
```
invalid blank line should be removed.

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 2]
```diff
@@ -64,6 +96,16 @@ func main() {
...
+	mp, err := initMeter()
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := mp.Stop(context.Background()); err != nil {
+			log.Printf("Error shutting down meter provider: %v", err)
+		}
+	}()
+
```
Comment (hanyuancheung) [inline instrumentation/net/http/otelhttp/example/server/server.go:108]:
```suggestion
```
Ditto.

**Review by hanyuancheung — APPROVED** [review]

File: instrumentation/net/http/otelhttp/example/go.mod [hunk 3]
```diff
@@ -7,15 +7,17 @@ replace go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp => ../
 require (
 	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp v0.33.0
 	go.opentelemetry.io/otel v1.8.1-0.20220713135543-5568a3072367
+	go.opentelemetry.io/otel/exporters/stdout/stdoutmetric v0.31.0
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/example/go.mod:10]: Please update this to use the latest metric SDK.

**Comment by fatsheep9146** [comment]
@dmathieu @MrAlias Already update to latest metric sdk, please help review this again, thanks!

File: CHANGELOG.md [hunk 4]
```diff
@@ -79,6 +79,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add gRPC trace `Filter` to the `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc` package to provide the way to filter the traces automatically generated in interceptors. (#2572)
 - The `TextMapPropagator` function to `go.opentelemetry.io/contrib/propagators/autoprop`.
   This function is used to return a composite `TextMapPropagator` from registered names (instead of having to specify with an environment variable). (#2593)
+- Enhance otelhttp example to support metric. (#2610)
```
Comment (MrAlias) [inline CHANGELOG.md:82]: Please move to the unreleased section.

Comment (MrAlias) [inline CHANGELOG.md:82]:
```suggestion
- Add example use of the metrics signal to `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp/example`. (#2610)
```

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 5]
```diff
@@ -16,27 +16,32 @@ package main
...
-	stdout "go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	stdoutmetric "go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
+	stdouttrace "go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
 	"go.opentelemetry.io/otel/propagation"
+
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
 	"go.opentelemetry.io/otel/sdk/resource"
 	sdktrace "go.opentelemetry.io/otel/sdk/trace"
-	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
+	semconv "go.opentelemetry.io/otel/semconv/v1.10.0"
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/example/server/server.go:37]:
```suggestion
	semconv "go.opentelemetry.io/otel/semconv/v1.12.0"
```

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 6]
```diff
@@ -16,27 +16,32 @@ package main
...
+	"os"
 
 	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
 
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/baggage"
-	stdout "go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	stdoutmetric "go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
+	stdouttrace "go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/example/server/server.go:31]:
```suggestion
	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
```

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 7]
```diff
@@ -53,6 +58,17 @@ func initTracer() (*sdktrace.TracerProvider, error) {
 	return tp, err
 }
 
+func initMeter() (*sdkmetric.MeterProvider, error) {
+	enc := json.NewEncoder(os.Stdout)
+	exp, err := stdoutmetric.New(stdoutmetric.WithEncoder(enc))
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/example/server/server.go:63]: Why not just use the default tab separated encoder the exporter uses?

```suggestion
	exp, err := stdoutmetric.New()
```

File: instrumentation/net/http/otelhttp/example/server/server.go [hunk 8]
```diff
@@ -53,6 +58,17 @@ func initTracer() (*sdktrace.TracerProvider, error) {
...
+	if err != nil {
+		return nil, err
+	}
+
+	mp := sdkmetric.NewMeterProvider(sdkmetric.WithReader(sdkmetric.NewPeriodicReader(exp)))
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/example/server/server.go:68]: This needs to be [set as the global meter provider](https://pkg.go.dev/go.opentelemetry.io/otel/metric@v0.33.0/internal/global#SetMeterProvider) or in the `main` loop in needs to be passed directly to the instrumentation.

It doesn't look like this is currently used. Have you tested this example to produce the telemetry you are expecting?

Comment (fatsheep9146) [reply instrumentation/net/http/otelhttp/example/server/server.go:66]: Sorry, I have tested in the first time, and saw the result is , so I thought the demo is right.
```
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:server"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.11.1"}}],"ScopeMetrics":[]}
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:server"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.11.1"}}],"ScopeMetrics":[]}
```
after setting the global meterprovider the result is 
```
{"Resource":[{"Key":"service.name","Value":{"Type":"STRING","Value":"unknown_service:server"}},{"Key":"telemetry.sdk.language","Value":{"Type":"STRING","Value":"go"}},{"Key":"telemetry.sdk.name","Value":{"Type":"STRING","Value":"opentelemetry"}},{"Key":"telemetry.sdk.version","Value":{"Type":"STRING","Value":"1.11.1"}}],"ScopeMetrics":[{"Scope":{"Name":"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp","Version":"semver:0.36.4","SchemaURL":""},"Metrics":[{"Name":"http.server.request_content_length","Description":"","Unit":"","Data":{"DataPoints":[{"Attributes":[{"Key":"http.flavor","Value":{"Type":"STRING","Value":"1.1"}},{"Key":"http.host","Value":{"Type":"STRING","Value":"http-server:7777"}},{"Key":"http.method","Value":{"Type":"STRING","Value":"GET"}},{"Key":"http.scheme","Value":{"Type":"STRING","Value":"http"}},{"Key":"http.server_name","Value":{"Type":"STRING","Value":"Hello"}}],"StartTime":"2022-11-05T04:39:46.704318664Z","Time":"2022-11-05T04:40:46.669193115Z","Value":0}],"Temporality":"CumulativeTemporality","IsMonotonic":true}},{"Name":"http.server.response_content_length","Description":"","Unit":"","Data":{"DataPoints":[{"Attributes":[{"Key":"http.flavor","Value":{"Type":"STRING","Value":"1.1"}},{"Key":"http.host","Value":{"Type":"STRING","Value":"http-server:7777"}},{"Key":"http.method","Value":{"Type":"STRING","Value":"GET"}},{"Key":"http.scheme","Value":{"Type":"STRING","Value":"http"}},{"Key":"http.server_name","Value":{"Type":"STRING","Value":"Hello"}}],"StartTime":"2022-11-05T04:39:46.704326714Z","Time":"2022-11-05T04:40:46.669195185Z","Value":14}],"Temporality":"CumulativeTemporality","IsMonotonic":true}},{"Name":"http.server.duration","Description":"","Unit":"","Data":{"DataPoints":[{"Attributes":[{"Key":"http.flavor","Value":{"Type":"STRING","Value":"1.1"}},{"Key":"http.host","Value":{"Type":"STRING","Value":"http-server:7777"}},{"Key":"http.method","Value":{"Type":"STRING","Value":"GET"}},{"Key":"http.scheme","Value":{"Type":"STRING","Value":"http"}},{"Key":"http.server_name","Value":{"Type":"STRING","Value":"Hello"}}],"StartTime":"2022-11-05T04:39:46.704333789Z","Time":"2022-11-05T04:40:46.669196421Z","Count":1,"Bounds":[0,5,10,25,50,75,100,250,500,750,1000,2500,5000,7500,10000],"BucketCounts":[0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"Min":0.102534,"Max":0.102534,"Sum":0.102534}],"Temporality":"CumulativeTemporality"}}]}]}
```

**Comment by fatsheep9146** [comment]
All comments are resolved @MrAlias

File: instrumentation/net/http/otelhttp/example/.vscode/settings.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+    "go.inferGopath": false
```
Comment (dmathieu) [inline instrumentation/net/http/otelhttp/example/.vscode/settings.json:2]: I don't think this file should be checked in.

**Comment by fatsheep9146** [comment]
the failed check seems to be flaky, could you trigger this again? @dmathieu @MrAlias

**Review by dmathieu — APPROVED** [review]

**Pull request merged**